File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.js
function solution(input) {
  const N = Number(input[0]);
  const map = input.slice(1).map((line) => line.split("").map(Number));
  const visited = Array.from({ length: N }, () => Array(N).fill(false));
  const dx = [-1, 1, 0, 0];
  const dy = [0, 0, -1, 1];

  const countList = [];

  function bfs(x, y) {
    const queue = [[x, y]];
    visited[x][y] = true;
    let count = 1;

    while (queue.length > 0) {
      const [cx, cy] = queue.shift();

      for (let dir = 0; dir < 4; dir++) {
        const nx = cx + dx[dir];
        const ny = cy + dy[dir];

        if (
          nx >= 0 &&
          nx < N &&
          ny >= 0 &&
          ny < N &&
          !visited[nx][ny] &&
          map[nx][ny] === 1
        ) {
          visited[nx][ny] = true;
          queue.push([nx, ny]);
          count++;
        }
      }
    }

    return count;
  }

  for (let i = 0; i < N; i++) {
    for (let j = 0; j < N; j++) {
      if (!visited[i][j] && map[i][j] === 1) {
        countList.push(bfs(i, j));
      }
    }
  }

  countList.sort((a, b) => a - b);
  console.log(countList.length);
  countList.forEach((count) => console.log(count));
}

const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");
solution(input);

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py

from collections import deque
import sys
input = sys.stdin.readline

def solution():
  N = int(input())
  maps = [list(map(int, input().strip())) for _ in range(N)]
  visited = [[False] * N for _ in range(N)]
  moves=[[-1, 0], [1, 0], [0, -1], [0, 1]]
  count_list = []
  
  def bfs(start):
    q = deque()
    q.append(start)
    visited[start[0]][start[1]] = True
    count = 1
    
    while q:
      x, y = q.popleft()
      
      for dx, dy in moves:
        nx, ny = x+dx, y+dy
        
        if nx < 0 or nx >= N or ny < 0 or ny >= N:
          continue
        if visited[nx][ny] or maps[nx][ny] == 0:
          continue
        
        visited[nx][ny] = True
        q.append([nx, ny])
        count += 1
        
    return count
  
  for i in range(N):
    for j in range(N):
      if not visited[i][j] and maps[i][j] == 1:
        count_list.append(bfs([i, j]))
        
  print(len(count_list))
  for c in sorted(count_list):
    print(c)
      

solution()